clash-lib/src/Clash/Core/Evaluator.hs

```diff
@@ -46,519 +44,359 @@ import           Clash.Core.Util
-  :: TyConMap
-  -> Heap
-  -> [Either Term Type]
-  -> (Heap,[Either Term Type])
-newLetBindings' tcm =
-    (second (map (either (Left . toVar) (Right . id))) .) . mapAccumL go
-  where
-    go h (Left tm)  = second Left (newLetBinding tcm h tm)
-    go h (Right ty) = (h,Right ty)
-
-mkAbstr
-  :: (Heap,Term)
-  -> [Either TyVar Type]
-  -> (Heap,Term)
-mkAbstr = foldr go
-  where
-    go (Left tv)  (h,e)          =
-      (h,TyLam tv (TyApp e (VarTy tv)))
-    go (Right ty) (Heap gh gbl h ids is,e) =
-      let ((ids',_),id_) = mkUniqSystemId (ids,is) ("x",ty)
-      in  (Heap gh gbl h ids' is,Lam id_ (App e (Var id_)))
-
--- | Force the evaluation of a variable.
-force :: Heap -> Stack -> Id -> Maybe State
-force (Heap gh g@(GPureHeap gbl) h ids is) k x' = case lookupVarEnv x' h of
-    Nothing -> case lookupVarEnv x' gbl of
-      Just e | isGlobalId x'
-        -> let e' = tickExpr e
-            in Just ( Heap gh (GPureHeap (delVarEnv gbl x')) h ids is
-                , GUpdate x':k
-                , deShadowTerm is e'
-                )
-      _ -> Nothing
-    Just e -> let e' = tickExpr e
-               in Just (Heap gh g (delVarEnv h x') ids is,Update x':k,e')
-    -- Removing the heap-bound value on a force ensures we do not get stuck on
-    -- expressions such as: "let x = x in x"
+  | otherwise
+  = let m' = heapInsert LocalId id_ e m
+     in (m' { mSupply = ids', mScopeNames = is1 }, id_)
  where
-  tickExpr = Tick (NameMod PrefixName (LitTy . SymTy $ toStr x'))
-  unQualName = snd . Text.breakOnEnd "."
-  toStr = Text.unpack . unQualName . flip Text.snoc '_' . nameOcc . varName
+  ty = termType tcm e
+  ((ids', is1), id_) = mkUniqSystemId (mSupply m, mScopeNames m) ("x", ty)
 
 -- | Unwind the stack by 1
 unwind
-  :: PrimEvaluator
-  -> TyConMap
-  -> Heap -> Stack -> Value -> Maybe State
-unwind eval tcm h k v = do
-  (kf,k') <- uncons k
-  case kf of
-    Update x                     -> return (update h k' x v)
-    GUpdate x                    -> return (gupdate h k' x v)
-    Apply x                      -> return (apply  h k' v x)
-    Instantiate ty               -> return (instantiate h k' v ty)
-    PrimApply nm ty tys vals tms -> primop eval tcm h k' nm ty tys vals v tms
-    Scrutinise _ alts            -> return (scrutinise h k' v alts)
-    -- Adding back the Tick constructor will make the evaluator loop
-    Tickish _                    -> return (h,k',valToTerm v)
-
--- | Update the Heap with the evaluated term
-update :: Heap -> Stack -> Id -> Value -> State
-update (Heap gh gbl h ids is) k x v = (Heap gh gbl (extendVarEnv x v' h) ids is,k,v')
-  where
-    v' = valToTerm v
-
--- | Update the Globals with the evaluated term
-gupdate :: Heap -> Stack -> Id -> Value -> State
-gupdate (Heap gh (GPureHeap gbl) h ids is) k x v =
-  (Heap gh (GPureHeap (extendVarEnv x v' gbl)) h ids is,k,v')
+  :: TyConMap
+  -> Machine
+  -> Value
+  -> Maybe Machine
+unwind tcm m v =
+  stackPop m >>= uncurry (flip go)
```

Comment:
I feel like point-free constructs like this are much harder to read than just writing it out with `do` notation.

Reply:
Yeah, I figured this might be near the edge of what you would accept
But on the other hand, it saves a whole one line :wink: 

```haskell
do
  (m', sf) <- stackPop m
  go sf m'
```